{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Guide RNA Clustering Homework\n",
    "## Objectives\n",
    "The goal of this homework was three fold:\n",
    "1. Analyze a set of sgRNA using a type of machine learning algorithm known as clustering in order to find interesting properties,\n",
    "2. Correlate these clusters with the activity of each guide to find a possible predictor of efficiency,\n",
    "3. Compute the genes that appear to be essential and find a possible explanation.\n",
    "\n",
    "## Clustering\n",
    "### Choosing features\n",
    "The first step of this work was to choose how the guide were going to be characterized.\n",
    "Eleven features were selected as a basis for the cluster analysis:\n",
    "* The number of nucleotides that make up the guide (A, T, G, C),\n",
    "* The count of several dinucleotide : GG, TA, TT, AG, AC and GC ; chosen for their importance in a previous work (Doench, 2016),\n",
    "* The melting temperature using a merged method based on three different thermodynamic tables (Panjkovich, 2005).\n",
    "\n",
    "### Computing features\n",
    "The count of nucleotides and dinucleotides for each guide was done using a script written in Python.\n",
    "The melting temperature was calculated using a software called [dnaMATE](melolab.org/dnaMATE/).\n",
    "A text file containing the features for each guide was used for the clustering.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import csv\n",
    "import numpy as np\n",
    "\n",
    "sequence_import = []\n",
    "tm_import = []\n",
    "sequence = []\n",
    "tm = []\n",
    "feature = []\n",
    "\n",
    "\n",
    "# Open files containing sgRNA and transfer them into an array\n",
    "with open('../data/RNA.txt', newline='') as inputfile:\n",
    "    for row in csv.reader(inputfile):\n",
    "        sequence_import.append(row)\n",
    "sequence = np.asarray(\n",
    "    [item for sublist in sequence_import for item in sublist])\n",
    "\n",
    "# Open files containing melting temperature and transfer them into an array\n",
    "with open('../data/Tm.txt', newline='') as inputfile:\n",
    "    for row in csv.reader(inputfile):\n",
    "        tm_import.append(row)\n",
    "tm = [item for sublist in tm_import for item in sublist]\n",
    "\n",
    "# Compute nucleotides feature for each sgRNA\n",
    "feature = np.zeros((len(sequence), 11))\n",
    "for idx, guide in enumerate(sequence):\n",
    "    feature[idx, 0] = guide.count('A')\n",
    "    feature[idx, 1] = guide.count('T')\n",
    "    feature[idx, 2] = guide.count('G')\n",
    "    feature[idx, 3] = guide.count('C')\n",
    "    feature[idx, 4] = guide.count('GG')\n",
    "    feature[idx, 5] = guide.count('TA')\n",
    "    feature[idx, 6] = guide.count('TT')\n",
    "    feature[idx, 7] = guide.count('AG')\n",
    "    feature[idx, 8] = guide.count('AC')\n",
    "    feature[idx, 9] = guide.count('GC')\n",
    "\n",
    "# Add melting temperature to the feature array\n",
    "for idx, temp in enumerate(tm):\n",
    "    feature[idx, 10] = temp\n",
    "\n",
    "np.savetxt('../data/feature.txt', feature)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clustering\n",
    "The clustering was done using the Scikit-learn machine learning library written in Python. The clustering was done using the k-means algorithm.\n",
    "A Python script was written in order to load the required data, compute the k-means and output the coordinates of each cluster as well as figures displaying the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from sklearn.cluster import KMeans\n",
    "\n",
    "import collections\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.cm as cm\n",
    "\n",
    "# Import features values and names from text files\n",
    "X = np.loadtxt(open(\"../data/feature.txt\"))\n",
    "y = np.asarray(open(\"../data/feature_list.txt\").readlines())\n",
    "\n",
    "# kmeans algorithm using scikit-learn\n",
    "n_clusters = 4\n",
    "clusterer = KMeans(init='k-means++', n_clusters=n_clusters, n_init=100)\n",
    "cluster_labels = clusterer.fit_predict(X)\n",
    "\n",
    "# Labeling the clusters\n",
    "centers = clusterer.cluster_centers_\n",
    "\n",
    "# Number of gene / cluster\n",
    "counter = collections.Counter(cluster_labels)\n",
    "distrib = []\n",
    "for i in range(n_clusters):\n",
    "    distrib.append(counter[i])\n",
    "\n",
    "# Export results to text files\n",
    "np.savetxt('../data/cluster_labels.txt', cluster_labels)  # clusters label for each RNA\n",
    "np.savetxt('../data/cluster.txt', centers)  # clusters coordinate\n",
    "np.savetxt('../data/distribution.txt', distrib)  # cluster distribution among guides\n",
    "\n",
    "# coordinate for average guide\n",
    "average_X = []\n",
    "for i in range(11):\n",
    "    average_X.append(np.average(X[:, i]))\n",
    "\n",
    "# Set colors for plots\n",
    "colors = cm.Paired(cluster_labels / n_clusters)\n",
    "\n",
    "# Plot between feature number 0 (A) and number 2 (G)\n",
    "plt.scatter(X[:, 0], X[:, 2], marker='.', s=500, lw=0, alpha=0.7, c=colors)  # Plot guides \n",
    "plt.xlabel(y[0])\n",
    "plt.ylabel(y[2])\n",
    "plt.scatter(centers[:, 0], centers[:, 2],\n",
    "            marker='o', c='white', alpha=1, s=500)  # Plot clusters\n",
    "for i, c in enumerate(centers):\n",
    "    plt.scatter(c[0], c[2], marker='$%d$' % i, alpha=1, s=50)\n",
    "plt.scatter(average_X[0], average_X[2], marker='x', c='black', alpha=1, s=300)  # Plot average guide\n",
    "plt.savefig('../images/0-2.png')\n",
    "plt.clf()\n",
    "\n",
    "# Plot between feature number 1 (T) and number 3 (C)\n",
    "plt.scatter(X[:, 1], X[:, 3], marker='.', s=500, lw=0, alpha=0.7, c=colors)\n",
    "plt.xlabel(y[1])\n",
    "plt.ylabel(y[3])\n",
    "plt.scatter(centers[:, 1], centers[:, 3],\n",
    "            marker='o', c='white', alpha=1, s=500)\n",
    "for i, c in enumerate(centers):\n",
    "    plt.scatter(c[1], c[3], marker='$%d$' % i, alpha=1, s=50)\n",
    "plt.scatter(average_X[1], average_X[3], marker='x', c='black', alpha=1, s=300)\n",
    "plt.savefig('../images/1-3.png')\n",
    "plt.clf()\n",
    "\n",
    "# Plot between feature number 4 (GG) and number 9 (GC)\n",
    "plt.scatter(X[:, 4], X[:, 9], marker='.', s=500, lw=0, alpha=0.7, c=colors)\n",
    "plt.xlabel(y[4])\n",
    "plt.ylabel(y[9])\n",
    "plt.scatter(centers[:, 4], centers[:, 9],\n",
    "            marker='o', c='white', alpha=1, s=500)\n",
    "for i, c in enumerate(centers):\n",
    "    plt.scatter(c[4], c[9], marker='$%d$' % i, alpha=1, s=50)\n",
    "plt.scatter(average_X[4], average_X[9], marker='x', c='black', alpha=1, s=300)\n",
    "plt.savefig('../images/4-9.png')\n",
    "plt.clf()\n",
    "\n",
    "# Plot between feature number 5 (TA) and number 10 (Tm)\n",
    "plt.scatter(X[:, 5], X[:, 10], marker='.', s=500, lw=0, alpha=0.7, c=colors)\n",
    "plt.xlabel(y[5])\n",
    "plt.ylabel(y[10])\n",
    "plt.scatter(centers[:, 5], centers[:, 10],\n",
    "            marker='o', c='white', alpha=1, s=500)\n",
    "for i, c in enumerate(centers):\n",
    "    plt.scatter(c[5], c[10], marker='$%d$' % i, alpha=1, s=50)\n",
    "plt.scatter(average_X[5], average_X[10], marker='x', c='black', alpha=1, s=300)\n",
    "plt.savefig('../images/5-10.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Results\n",
    "Several parameters for the k-means algorithm were tested like the number of cluster, the reduction of the data using PCA, the method for initializing the cluster.\n",
    "The chosen parameters were the following:\n",
    "* Number of cluster = 4,\n",
    "* No dimensionality reduction,\n",
    "* Number of initialization = 100.\n",
    "The resulting coordinates for the 4 clusters are showed in table 1.\n",
    "\n",
    "![table](images/Clusters coordinate.png)\n",
    "\n",
    "The main characteristics on average for each cluster are as follows:\n",
    "* Cluster 0 is made of guides with a low amount of GC and GG but with a high amount of TA and a low\n",
    "Tm as well as a high amount of T and a medium amount of A.\n",
    "* Cluster 1 is made of guides with a high amount of GC and GG but with a low amount of TA and a high\n",
    "Tm.\n",
    "* Cluster 2 is made of guides similar to the one of Cluster 0 (low GC and GG) but have a medium amount\n",
    "of T and a high amount of A.\n",
    "* Cluster 3 is made of guides that have medium value in all the features.\n",
    "Concerning the amount of guides per cluster, we can see that cluster number 0 to 2 have roughly the same\n",
    "number (from 20 to 23% of the sgRNA) while cluster 3 (the ‘medium’ cluster) encapsulate more than a 3rd of\n",
    "the guide.\n",
    "\n",
    "Several figures were created to visualize the difference between each cluster and of far they are for the\n",
    "average “guide” (black cross on each plot). The following color were attributed to each cluster: cluster 0 in\n",
    "blue, cluster 1 in green, cluster 2 in orange and cluster 3 in violet. The center for each cluster was represented\n",
    "using a circle filled the number of the cluster.\n",
    "![plots](images/plots.png)\n",
    "The top plots allowed to visualize the difference between the first 3 clusters while it is hard to see any area\n",
    "belonging to cluster 3.\n",
    "\n",
    "The bottom plots allowed to visualize the position occupied by cluster 3 with a medium TM and the similarity between\n",
    "cluster 0 and 2.\n",
    "## Predicting activity\n",
    "### Guide activity\n",
    "The next step was to determine the activity of each guide. This was done using the following equation:\n",
    "![activity](images/activity.png)\n",
    "With:\n",
    "* count t2 = average normalized count of cells with the drug PLX after 14 days,\n",
    "* count t1 = normalized count of cells before introduction of the plasmid.\n",
    "The activity was then normalized by percent-rank as described in another paper (Doench, 2014).\n",
    "Analysis\n",
    "The relation between each guide’s cluster and percent-rank was done using a Python script that divided the\n",
    "guides in quartile based on their percent-rank and then assigned them to their corresponding cluster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "# Import activity and cluster assignement for each guide\n",
    "rank = np.loadtxt('../data/rank.txt', delimiter=\"\\t\")\n",
    "\n",
    "# Divide guides based on their activity and cluster\n",
    "quartile = np.zeros(shape=(4, 4))\n",
    "for q in range(4):\n",
    "    for c in range(4):\n",
    "        count = 0\n",
    "        for g in range(len(rank)):\n",
    "            cluster = rank[g][1]\n",
    "            activity = rank[g][0]\n",
    "            if cluster == c and activity < (25.0 + q * 25.0) and activity >= (q * 25.0):\n",
    "                count = count + 1\n",
    "            if q == 4 and cluster == c and activity == 100:\n",
    "                count = count + 1\n",
    "            if q == 0 and cluster == c and activity == 0:\n",
    "                count = count + 1\n",
    "        quartile[c][q] = count\n",
    "\n",
    "# Export results as a text file\n",
    "np.savetxt('../data/prediction.txt', quartile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Table 2 display the resulting distribution.\n",
    "\n",
    "![distribution](images/distribution.png)\n",
    "\n",
    "We observe that cluster 0 has an even distribution between each quartile, which would not make it a good\n",
    "predictor of activity.\n",
    "Cluster 1 is the opposite of a good predictor with only 20% of its guide having a rank of 75% or higher but 29%\n",
    "with a rank of 25% of lower.\n",
    "Cluster 3 has a distribution similar to cluster 1 with only 22% of its guides having a rank of 75% or higher.\n",
    "Finally, cluster 2 seems to be a good predictor of activity with 34% of its guides having at least a rank of 75%.\n",
    "\n",
    "## Guide analysis\n",
    "The average activity of each gene was determined and a percent-rank was created.\n",
    "The TOP5 genes that had the most impact when targeted were: NHP2L1, NIP7, RPS18, DDX56 and RPL8 (table\n",
    "3).\n",
    "\n",
    "![gene](images/gene.png)\n",
    "\n",
    "Using NCBI RefSeq and UniProt, the following information were extracted for each gene:\n",
    "* NHP2L1: Nuclear protein that binds snRNA and may play a role in late stage of spliceosome assembly,\n",
    "* NIP7: Protein required for proper 34S pre-rRNA processing and 60S ribosome subunit assembly,\n",
    "* RPS18: Ribosomal protein S18, located at the top of the head of the 40S subunit,\n",
    "* DDX56: DEAD box proteins that may play a role in processing of pre-ribosomal particles that lead to\n",
    "mature 60S ribosomal subunits,\n",
    "* RPL8: Ribosomal protein L8, part of the 60S subunit.\n",
    "We observe that each of these genes play a major role in the cells vital mechanisms.\n",
    "\n",
    "## Conclusion\n",
    "The clustering algorithm divided the guides into 4 clusters mainly by using the Tm, the amount of A and T\n",
    "nucleotide.\n",
    "Using the activity of each guide, we concluded that guides with a low amount of GC and GG but with a medium\n",
    "amount of T and a high amount of A were more susceptible to have a high activity.\n",
    "Finally, we observed that the most depleted guides were targeting genes essential for the cells survival.\n",
    "\n",
    "## Discussion\n",
    "The goal of this work was just to provide a basis and can be improved a lot.\n",
    "The chosen feature were small in number and simple. It would be interesting to add more features based on\n",
    "the energy and entropy of this system. It would also be interesting not to look only at the guide but also at its\n",
    "surrounding environment and target.\n",
    "The number of clusters was intentionally kept low to decrease the computing time and facilitate the\n",
    "interpretation. The algorithm used was only one among many and it would be interesting to compare it to\n",
    "others and more complex ones.\n",
    "The activity was only described as the log2 between t0 and 14 days after with the drug. It would be interesting\n",
    "to compute a measure of how fast we see change by using both value at 7 and 14 days, with or without drug.\n",
    "Finally, it would be interesting to have more information on the targets to allow further discrimination\n",
    "between the genes (essential versus non-essential).\n",
    "\n",
    "## Bibliography\n",
    "*Doench, John G., et al. \"Rational design of highly active sgRNAs for CRISPR-Cas9-mediated gene inactivation.\"\n",
    "Nature biotechnology 32.12 (2014): 1262-1267.*\n",
    "\n",
    "*Doench, John G., et al. \"Optimized sgRNA design to maximize activity and minimize off-target effects of\n",
    "CRISPR-Cas9.\" Nature biotechnology (2016).*\n",
    "\n",
    "*Panjkovich, Alejandro, and Francisco Melo. \"Comparison of different melting temperature calculation\n",
    "methods for short DNA sequences.\" Bioinformatics 21.6 (2005): 711-722.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
